pkg/kubelet/pending_and_terminating_handler.go

```diff
@@ -0,0 +1,89 @@
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package kubelet
+
+import (
+	"fmt"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/klog/v2"
+	"k8s.io/kubernetes/pkg/kubelet/lifecycle"
+	"k8s.io/kubernetes/pkg/kubelet/status"
+)
+
+const (
+	deletedWhilePendingReason  = "Deleted"
+	deletedWhilePendingMessage = "Deleted while in the pending phase"
+)
+
+// pendingAndTerminatingHandler determines pending and terminating pods to be evicted
+type pendingAndTerminatingHandler struct {
+	// the provider of pod status
+	podStatusProvider status.PodStatusProvider
+	// the recorder to dispatch events when we identify a pod to be evicted
+	recorder record.EventRecorder
+}
+
+// newPendingAndTerminatingHandler
+func newPendingAndTerminatingHandler(
+	podStatusProvider status.PodStatusProvider,
+	recorder record.EventRecorder,
+) (*pendingAndTerminatingHandler, error) {
+	// check for all required fields
+	if podStatusProvider == nil {
+		return nil, fmt.Errorf("pod status provider must not be nil")
+	}
+	if recorder == nil {
+		return nil, fmt.Errorf("event recorder must not be nil")
+	}
+	return &pendingAndTerminatingHandler{
+		podStatusProvider: podStatusProvider,
+		recorder:          recorder,
+	}, nil
+}
+
+// ShouldSync returns true if the pod should be evicted.
+func (m *pendingAndTerminatingHandler) ShouldSync(pod *v1.Pod) bool {
+	return m.shouldBeEvicted(pod)
+}
+
+// ShouldEvict returns true if the pod should be evicted.
+func (m *pendingAndTerminatingHandler) ShouldEvict(pod *v1.Pod) lifecycle.ShouldEvictResponse {
+	if !m.shouldBeEvicted(pod) {
+		return lifecycle.ShouldEvictResponse{Evict: false}
+	}
+	klog.V(5).InfoS("Deleting a pending and terminating pod", "pod", klog.KObj(pod))
+	m.recorder.Eventf(pod, v1.EventTypeNormal, deletedWhilePendingReason, deletedWhilePendingMessage)
+	return lifecycle.ShouldEvictResponse{Evict: true, Reason: deletedWhilePendingReason, Message: deletedWhilePendingMessage}
+}
+
+// shouldBeEvicted determines if a pod should be evicted by the handler
+func (m *pendingAndTerminatingHandler) shouldBeEvicted(pod *v1.Pod) bool {
+	// this handler only evicts pods which are Terminating
+	if pod.DeletionTimestamp == nil {
+		return false
+	}
+	// get the latest status to check the phase
+	podStatus, found := m.podStatusProvider.GetPodStatus(pod.UID)
+	if !found {
+		// fallback to the status from the api server. This is needed for situations
+		// like Kubelet restart.
+		podStatus = pod.Status
+	}
```

Comment:
I'm curious why this case was needed, during kubelet restart, what is the scenario of kubelet restart that requires this case?

Reply:
The handler does the lookup to the `kl.statusManager` which is empty after kubelet restart and gets populated by `SetPodStatus` , called after `generateAPIPodStatus`. An analogous fallback is also done inside `generateAPIPodStatus`: https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubelet_pods.go#L1488-L1491.

Doing the fallback in yet another place might be another argument for dropping the handler idea, and moving the state explicitly in `generateAPIPodStatus`. WDYT?

Reply:
Yes, I think it would be clearer (and less changes) to move this to `generateAPIPodStatus`. It also keeps all the main api pod transitions in the same function which I think is ideal.

Reply:
Actually, there are transitions outside of this function, which are ecictions, for example this: https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/eviction/eviction_manager.go#L584.

Reply:
So, I'm wondering when my case really belongs best inside `generateAPIPodStatus`. I think `getPhase` isn't ideal, because its't intention seems to be to set the phase according to the current status of containers. In my case, the containers might still be running.

Reply:
Moved the code (but keep the old version on a backup branch just in case :)). PTAL.